urld829b80034_wjch111_PTSP2013Wang_tgz-pJ8-controllers_momcts_momctsCore_TestPlannerJ8/src/framework/core/Controller.java

```diff
@@ -0,0 +1,286 @@
+     * Action that involves thrust and right rotation.
+     */
+    public static final int ACTION_THR_RIGHT = 5;
+
+    /**
+     * Precomputed constant for PI/2.
+     */
+    public static final double HALF_PI = Math.PI * 0.5;
+
+    /**
+     * Precomputed constant for PI/4.
+     */
+    public static final double QUARTER_PI = Math.PI * 0.25;
+
+    /**
+     * Number of actions for the ship.
+     */
+    public static final int NUM_ACTIONS = 6;
+
+    /**
+     * Indicates if the thread of the controller is alive.
+     */
+    private boolean m_alive;
+
+    /**
+     * Indicates if the controller has received notification of another game cycle.
+     */
+    private boolean m_wasSignalled;
+
+    /**
+     * Indicates if the controller has computed an action for the next cycle.
+     */
+    private boolean m_hasComputed;
+
+    /**
+     * Indicates where the controller is due to execute an action.
+     */
+    private long m_timeDue;
+
+    /**
+     * Reference to the game.
+     */
+    private Game m_game;
+
+    /**
+     * Last move executed by the controller.
+     */
+    private int m_lastMove;
+
+    /**
+     * Tick of the game when a move was requested.
+     */
+    private int m_tick;
+
+    /**
+     * Instantiates a new controller. The constructor initialises the class variables.
+     */
+    public Controller()
+    {
+        m_alive=true;
+        m_wasSignalled=false;
+        m_hasComputed=false;
+        m_tick=0;
+    }
+
+    /**
+     * Resets the state of the controller.
+     */
+    public synchronized void reset()
+    {
+        m_alive=true;
+        m_wasSignalled=false;
+        m_hasComputed=false;
+    }
+
+    /**
+     * Terminates the controller: a signal is sent and the flag 'alive' is set to false. When
+     * the thread wakes up, the outer loop will terminate and the thread finishes.
+     */
+    public final synchronized void terminate()
+    {
+        m_alive=false;
+        m_wasSignalled=true;
+        notify();
+    }
+
+    /**
+     * Updates the game state: a copy of the game is passed to this method and the class variable is
+     * updated accordingly.
+     *
+     * @param a_game A copy of the current game
+     * @param a_timeDue The time the next move is due
+     */
+    public final synchronized void update(Game a_game,long a_timeDue)
+    {
+        //System.out.println("update");
+        this.m_game=a_game;
+        this.m_timeDue=a_timeDue;
+        m_wasSignalled=true;
+        m_hasComputed=false;
+        m_lastMove = Controller.ACTION_NO_FRONT;
+        notify();
+    }
+
+    /* (non-Javadoc)
+	 * @see java.lang.Runnable#run()
+	 */
+    public final void run()
+    {
+        while(m_alive)
+        {
+            synchronized(this)
+            {
+                while(!m_wasSignalled)
+                {
+                    try
+                    {
+                        wait();
+                    }
+                    catch(InterruptedException e)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)